wgengine/tsdns/tsdns.go

```diff
@@ -0,0 +1,269 @@
+package tsdns
+
+import (
+	"encoding/binary"
+	"errors"
+	"sync"
+
+	"github.com/tailscale/wireguard-go/device"
+	dns "golang.org/x/net/dns/dnsmessage"
+	"inet.af/netaddr"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/packet"
+)
+
+// MaxResponseSize is the maximal size of a Tailscale DNS response.
+const MaxResponseSize = 512
+
+const (
+	// ipOffset is the space before the IP header. It is reserved for wireguard-go.
+	ipOffset = device.MessageTransportHeaderSize
+	// dnsDataOffset is the space before DNS data. It includes the IP and UDP headers.
+	dnsDataOffset = ipOffset + packet.UDPDataOffset
+
+	// The response buffer must have space for all the headers and the response body.
+	bufferSize = dnsDataOffset + MaxResponseSize
+)
+
+var (
+	errMapNotSet      = errors.New("domain map not set")
+	errNoSuchDomain   = errors.New("domain does not exist")
+	errNotImplemented = errors.New("query type not implemented")
+	errNotOurName     = errors.New("not an *.ipn.dev domain")
+	errNotQuery       = errors.New("not a DNS query")
+)
+
+var (
+	// The default IP for a new resolver.
+	DefaultIP = packet.IP(binary.BigEndian.Uint32([]byte{100, 100, 100, 100}))
+	// The default port for a new resolver.
+	DefaultPort = uint16(53)
+)
+
+// DNSMap is all the data Resolver needs to resolve DNS queries.
+type DNSMap struct {
+	// DomainToIP is a mapping of Tailscale domains to their IP addresses.
+	// For example, monitoring.ipn.dev -> 100.64.0.1.
+	DomainToIP map[string]netaddr.IP
+}
+
+// Resolver is a DNS resolver for domain names of the form *.ipn.dev
+type Resolver struct {
+	logf logger.Logf
+
+	// ip is the IP on which the resolver is listening.
+	ip packet.IP
+	// port is the port on which the resolver is listening.
+	port uint16
+
+	// parser is a request parser that is reused by the resolver.
+	parser dns.Parser
+	// responseBuffer is a static buffer to avoid graticious allocations.
+	responseBuffer [bufferSize]byte
+
+	// mu guards the following fields from being updated while used.
+	mu sync.Mutex
+	// dnsMap is the map most recently received from the control server.
+	dnsMap *DNSMap
+}
+
+// NewResolver constructs a resolver with default parameters.
+func NewResolver(logf logger.Logf) *Resolver {
+	return &Resolver{
+		logf: logf,
+		ip:   DefaultIP,
+		port: DefaultPort,
+	}
+}
+
+// AcceptsPacket determines if the given packet is
+// directed to this resolver (by ip and port).
+// We also require that UDP be used to simplify things for now.
+func (r *Resolver) AcceptsPacket(in *packet.QDecode) bool {
+	return in.DstIP == r.ip && in.DstPort == r.port && in.IPProto == packet.UDP
+}
+
+// SetDNSMap sets the resolver's DNS map.
+func (r *Resolver) SetDNSMap(dm *DNSMap) {
+	r.mu.Lock()
+	r.dnsMap = dm
+	r.mu.Unlock()
+}
+
+// Resolve maps a given domain name to the IP address of the host that owns it.
+func (r *Resolver) Resolve(domain string) (netaddr.IP, dns.RCode, error) {
+	const suffixLength = len(".ipn.dev")
```

Comment:
can drop this and just do:

     if !strings.HasSuffix(domain, ".ipn.dev") {

     }

Should inline to the same code, but more readable.